crates/dns/src/kube.rs

```diff
@@ -0,0 +1,251 @@
+    /// The FQDN for the service. Typically of the form `<name>.<namespace>.svc.cluster.local`.
+    pub fqdn: Name,
+
+    /// The name of the service.
+    pub name: String,
+
+    /// The namespace of the service.
+    pub namespace: String,
+}
+
+/// Generates the set of valid alternative hosts for the service FQDN
+/// (e.g. `s1.ns1.svc.cluster.local`), based on the location of the client.
+///
+/// # Arguments
+///
+/// * `client` - Information about the Kubernetes client.
+/// * `service` - Information about the Kubernetes service being called from the client.
+///
+/// # Returns
+///
+/// If successful, returns a list of alternate hosts for the Kubernetes service. This list
+/// can include the following:
+///
+/// * `<fqdn>` - the original FQDN of the service that was passed in. This will always be
+///              included in the output.
+/// * `<name>.` - added only if the client and service reside in the same domain
+///               and namespace.
+/// * `<name>.<namespace>` - added only if the client and service reside in the same domain.
+/// * `<name>.<namespace>.<svc>` - added only if the client and service reside in the same
+///                                domain. The value of `svc` is taken as the first label
+///                                in the domain. Since the domain in Kubernetes is typically
+///                                `svc.cluster.local`, the value of the first label is
+///                                typically `svc`.
+///
+/// Returns any errors encountered during processing of domain labels.
+///
+pub fn gen_kube_alt_hosts(
+    client: &ClientInfo,
+    service: &ServiceInfo,
+) -> ProtoResult<Vec<LowerName>> {
+    let mut out = Vec::new();
+
+    if !client.domain.is_fqdn() {
+        return Err(ProtoError::from("client domain is not an FQDN"));
+    }
+
+    if !service.fqdn.is_fqdn() {
+        return Err(ProtoError::from("service FQDN is not an FQDN"));
+    }
+
+    out.push(service.fqdn.clone().into());
+
+    // Do not generate alt hostnames if the client and service reside in different domains.
+    if !client.domain.zone_of(&service.fqdn) {
+        return Ok(out);
+    }
+
+    let num_fqdn_labels = service.fqdn.num_labels();
+
+    // If the client and server reside in the same namespace, add an alias for just the
+    // service name 'name.'.
+    if num_fqdn_labels > 1 && client.namespace == service.namespace {
+        let name_label = service.fqdn.iter().next().unwrap();
+        out.push(Name::from_labels(vec![name_label])?.into());
+    }
+
+    // Add 'name.namespace.'.
+    if num_fqdn_labels > 2 {
+        let mut iter = service.fqdn.iter();
+        let name_label = iter.next().unwrap();
+        let ns_label = iter.next().unwrap();
+        out.push(Name::from_labels(vec![name_label, ns_label])?.into());
+    }
+
+    // TODO(nmittler): Do we need this?
+    // Do we need to generate entries for 'name.namespace.svc', 'name.namespace.svc.cluster', etc. ?
```

Comment:
`.svc` - yes

`.svc.cluster` is not a valid domain

"If these are not that frequently used," is not relevant IMO. Correctness >>>> everything else

Reply:
Yeah ... I copied this comment over from the original code. I'm fine with just getting rid of it.